Cycles Windows

ericmehl opened the pull request:
This gets Cycles building and running on Windows. Only the last three commits are relevant, there's a pile of unrelated commits at the start because it's based on #4869 and #4934.

Of particular interest in the last commit, where I made some changes to the scene locking behavior. I was getting a deadlock when doing any scene update where it was waiting for `m_scene->mutex` to be freed up at https://github.com/GafferHQ/gaffer/blob/9199b2d50bcc66d6b5f5b746e5e580b13910a1d6/src/GafferCycles/IECoreCyclesPreview/Renderer.cpp#L3147

I _think_ my logic is sound here because `std::scoped_lock` takes ownership of the mutex, so going out of scope won't unlock another scene's mutex if it changed in one of the `update*` methods. I don't know why this was only causing problems on Windows.

- List specific new features and changes to project components

### Related issues ###

- List any Issues this PR addresses or solves

### Dependencies ###

#4869 and #4934

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

boberfly commented:
LGTM - The code originally catered for global option changes that needed to do hard resets that involved changes to `ccl::Scene` or `ccl::Session` like the Blender translation code which means deleting it and re-creating it all, but I've given up on this and will just flag a warning that the renderer needs to be restarted manually so the complexity removal is welcomed here.

Having a scoped lock is much better here for sure.

johnhaddon commented:
> The code originally catered for global option changes that needed to do hard resets that involved changes to ccl::Scene or ccl::Session like the Blender translation code which means deleting it and re-creating it all, but I've given up on this

@boberfly, can we also remove `m_sessionReset` now? It only seems to get set true by `option( "cycles:device" )`, so I'm wondering if it is just the vestigial tail of this old hard reset code you mentioned.

johnhaddon commented:
Oh, and I think we can definitely get rid of `m_sceneChanged` can't we? It gets set in `option()` and `render()`, but never gets actually used.

ericmehl commented:
I rebased this onto `main` just now, and also enabled Cycles tests on Windows. I went ahead and removed `m_sessionReset` - it was being used in a couple of places, but I don't think it was actually resetting anything that wasn't already being reset.

Ready for a review

ericmehl commented:
Any chance you've come across the problem where Cycles can't find the CPU renderer, @boberfly? If not, I guess we could push a temporary commit that lists the available modes and see what the CI machines offer us.

johnhaddon commented on .github/workflows/main.yml:
```diff
@@ -66,7 +66,7 @@ jobs:
             dependenciesURL: https://github.com/ImageEngine/cortex/releases/download/10.4.3.0/cortex-10.4.3.0-windows-python3.zip
             testRunner: Invoke-Expression
-            testArguments: GafferTest GafferVDBTest GafferUSDTest
+            testArguments: GafferTest GafferVDBTest GafferUSDTest GafferCyclesTest
```
Oh, forgot to ask - does the device problem only occur on the CI workers? Or does it happen locally too?

ericmehl replied:
It only happens on CI workers. All of the tests pass fine on my local machine, which makes me feel better about leaving this out for the immediate future.

ericmehl commented:
I squashed in the two changes and dropped the commit adding `GafferCyclesTest`. Should be ready to go!
